### PR TITLE
Code analysis 3

### PR DESCRIPTION
**Pull request by ariccio**
Retry of #315

### CODE REVIEW

**Comment by janisozaur** [comment]
@ariccio why do you keep closing and submitting new PRs?

File: src/CalcManager/Ratpack/num.cpp [hunk 1]
```diff
@@ -192,9 +192,9 @@ void _addnum( PNUMBER *pa, PNUMBER b, uint32_t radix)
 //
 //----------------------------------------------------------------------------
 
-void _mulnum( PNUMBER *pa, PNUMBER b, uint32_t radix);
+void _mulnum( _Inout_ PNUMBER *pa, _In_ _Const_ PNUMBER b, uint32_t radix);
```
Comment (janisozaur) [inline src/CalcManager/Ratpack/num.cpp]: What's the purpose of `_Const_` here? What benefit does it bring over `const`?

File: src/CalcManager/Ratpack/rat.cpp [hunk 2]
```diff
@@ -99,8 +99,8 @@ void fracrat( PRAT *pa , uint32_t radix, int32_t precision)
 //
 //-----------------------------------------------------------------------------
 
-void mulrat( PRAT *pa, PRAT b, int32_t precision)
-
+void mulrat( _Inout_ PRAT *pa, _In_ PRAT b, int32_t precision)
+
```
Comment (janisozaur) [inline src/CalcManager/Ratpack/rat.cpp]:
```suggestion
```
Please don't introduce unnecessary whitespace

File: src/CalcManager/Ratpack/rat.cpp [hunk 3]
```diff
@@ -263,8 +263,8 @@ void addrat( PRAT *pa, PRAT b, int32_t precision)
 //
 //-----------------------------------------------------------------------------
 
-void rootrat( PRAT *py, PRAT n, uint32_t radix, int32_t precision)
-{
+void rootrat( _Inout_ PRAT *py, _In_ PRAT n, uint32_t radix, int32_t precision)
+{
```
Comment (janisozaur) [inline src/CalcManager/Ratpack/rat.cpp]:
```suggestion
{
```

File: src/CalcManager/Ratpack/conv.cpp [hunk 4]
```diff
@@ -373,7 +373,7 @@ PNUMBER nRadixxtonum( _In_ PNUMBER a, uint32_t radix, int32_t precision)
 //
 //-----------------------------------------------------------------------------
 
-PNUMBER numtonRadixx(_In_ PNUMBER a, uint32_t radix)
+PNUMBER numtonRadixx(PNUMBER a, uint32_t radix)
```
Comment (janisozaur) [inline src/CalcManager/Ratpack/conv.cpp]: I'm surprised by this change, was this intended?

File: src/CalcManager/Ratpack/num.cpp (hunk 1 above)
Comment (ariccio) [reply src/CalcManager/Ratpack/num.cpp]: I don't want to change the interface of _mulnum, and (IIRC) it does actually modify it a few times within the function, before returning it to the original state.

File: src/CalcManager/Ratpack/conv.cpp (hunk 4 above)
Comment (ariccio) [reply src/CalcManager/Ratpack/conv.cpp]: Honestly, I'm not 100% sure, but I think it was because `a` is actually modified.

Comment (janisozaur) [reply src/CalcManager/Ratpack/conv.cpp]: Wouldn't that make it `_Inout_`?

File: src/CalcManager/Ratpack/num.cpp (hunk 1 above)
Comment (janisozaur) [reply src/CalcManager/Ratpack/num.cpp]: That makes no sense. Either it is const, in which case it should be marked so with `const` or it isn't and you shouldn't try faking it.

File: src/CalcManager/Ratpack/exp.cpp [hunk 5]
```diff
@@ -226,8 +226,8 @@ void lograt( PRAT *px, int32_t precision)
     destroyrat(offset);
     destroyrat(pwr);
 }
-
-void log10rat( PRAT *px, int32_t precision)
+
```
Comment (janisozaur) [inline src/CalcManager/Ratpack/exp.cpp]:
```suggestion

```
Can't you just strip all the modified files?

File: src/CalcManager/Ratpack/num.cpp (hunk 1 above)
Comment (ariccio) [reply src/CalcManager/Ratpack/num.cpp]: I think you're correct here.

File: src/CalcManager/Ratpack/conv.cpp (hunk 4 above)
Comment (ariccio) [reply src/CalcManager/Ratpack/conv.cpp]: In this codebase, I think they intend to treat `PNUMBER` and such as conceptually `_In_` arguments, since they usually pass parameters that are conceptually modified as pointer-to-pointer (`PNUMBER*`) or reference (`PNUMBER&`). I'm going to recompile with this as an `_In_` argument and see if it complains.

Comment (ariccio) [reply src/CalcManager/Ratpack/conv.cpp]: Doesn't seem to complain, will return to original.

File: src/CalcViewModel/CalcViewModel.vcxproj [hunk 6]
```diff
@@ -40,6 +40,7 @@
     <RootNamespace>CalcViewModel</RootNamespace>
     <DefaultLanguage>en-US</DefaultLanguage>
     <MinimumVisualStudioVersion>14.0</MinimumVisualStudioVersion>
+    <PreferredToolArchitecture>x64</PreferredToolArchitecture>
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/CalcViewModel.vcxproj]: According to the [docs](https://docs.microsoft.com/en-us/cpp/build/walkthrough-using-msbuild-to-create-a-visual-cpp-project), this is supposed to appear _after_ the import of `Microsoft.Cpp.default.props` and here you have it before.  You can take a look at how CalcManager sets this up:
https://github.com/Microsoft/calculator/blob/ad25feda6bfac4b4779b92500b01d341931dcbb0/src/CalcManager/CalcManager.vcxproj#L51-L55

File: src/CalcViewModel/Common/Utils.h [hunk 7]
```diff
@@ -284,8 +284,8 @@ namespace Utils
 
     void IFTPlatformException(HRESULT hr);
     Platform::String^ GetStringValue(Platform::String^ input);
-    bool IsLastCharacterTarget(std::wstring const &input, wchar_t target);
-    std::wstring RemoveUnwantedCharsFromWstring(std::wstring inputString, wchar_t* unwantedChars, unsigned int size);
+    bool IsLastCharacterTarget(_In_ std::wstring const &input, wchar_t target);
```
Comment (danbelcher-MSFT) [inline src/CalcViewModel/Common/Utils.h]:
```suggestion
    bool IsLastCharacterTarget(std::wstring const &input, wchar_t target);
```

**Comment by danbelcher-MSFT** [comment]
Ping @ariccio

**Comment by ariccio** [comment]
Sorry, I can't really put a whole lot of time into this for the next week or so - I'm quite busy -

**Comment by danbelcher-MSFT** [comment]
No problem!  This improvement can come in when you're ready.

**Comment by mcooley** [comment]
Many of the improvements here made it into #689.

I'm closing out this PR since it's been inactive for a while, but I know the code here will be a useful reference if you or others want to enable/fix more code analysis warnings in the future. We're tracking that work as #693 now. Thanks for contributing to Calculator!
